Feat: Distribute users without rating

PhMemmel opened the pull request:
As followup for the group restriction PR I developed a feature targeting #232 and #124.

As stated in #232 by @danmarsden there is another solution for this which however seems to go for a different approach.

@danmarsden, @Laur0r Let me know how to deal with that. I can offer to integrate all 3 ways (the 2 ways in this PR and https://github.com/danmarsden/moodle-mod_ratingallocate/commit/e9c71d7612f6cfa021caf413e46273bb7f66ec3d) of distributing unallocated users? However, it looks to me as the feature of @danmarsden depends on some other features? So maybe this should be handled differently.

Laur0r commented on locallib.php:
```diff
@@ -621,6 +624,248 @@ private function process_action_manual_allocation() {
+        $choicessortedwithgroupscount = [];
+        $choicessorted = [];
+        foreach (array_keys($possiblechoices) as $choiceid) {
+            $choicessortedwithgroupscount[count($this->get_choice_groups($choiceid))][] = $choiceid;
```
It should be checked first whether there exists an active group restriction as the groups are not deleted directly when the restriction is disabled.

PhMemmel replied:
I'm not 100% sure, if I understand your point. But are you aware of lines 735-737:
```
            if (empty($choice->usegroups)) {
                // If we have a group without group restrictions it will always be available.
                continue;
            }
```

Laur0r replied:
Yes, but these lines do not change that these choices are still considered here. The problem is that whenever you have a choice with group restrictions e.g. for two groups and you change the restriction by disabling it, those two group restrictions are still in the database. So when `get_choice_groups` is called, this would return those two groups although this is actually a choice which does not have an active group restriction any longer. Consequently, those choices are handled differently for the allocation as they are sorted by their group count. Therefore, there should be a check first at this point if `empty($choice->usegroups)`

PhMemmel replied:
I again dove into the code and you are absolutely right. In my mind (and unfortunately only there ;-)) `get_choice_groups` returns an empty array in case the choice is not active, but that's obviously not the case. I think I also did not handle the case that a choice is disabled, right? I will have to sort this out as well in this algorithm?

PhMemmel replied:
Just realized, I did respect disabled choices, so everything should be fine. I fixed what you asked for. Very good catch! Thanks!

Laur0r reviewed (commented):
Thank you for your contribution and thoughtful comments! I added comments for some problems I encountered during testing. Otherwise the algorithms seem to work well.

PhMemmel commented on locallib.php:
```diff
@@ -835,6 +1080,16 @@ public function handle_view() {
+            case ACTION_DISTRIBUTE_UNALLOCATED_FILL:
+                $this->distribute_users_without_choice($action);
+                redirect(new moodle_url('/mod/ratingallocate/view.php',
+                            ['id' => $this->coursemodule->id, 'action' => ACTION_NONE]),
```
No idea why this is there, I removed the action parameter completely

PhMemmel reviewed (commented):
Thank you very much for the review and the issues you found. I tried to address all of them. Any more feedback?

PhMemmel commented:
I also rebased and fixed the issue why tests failed on current master.

Laur0r commented on locallib.php:
```diff
@@ -621,6 +624,248 @@ private function process_action_manual_allocation() {
+    public function get_undistributed_users(): array {
+        $undistributedusers = [];
+        $userswithgroups = $this->get_undistributed_users_with_groupscount();
+        for ($i = 1; $i <= max(array_keys($userswithgroups)); $i++) {
```
I understand your hesitation. I tested it for our larger courses with over 500 students and there didn't seem to be too much of a slow down.

Laur0r reviewed (commented):
Thank you! Two more considerations that have to be made:
1. When distributing a large number of students, this may require more memory or time than is initially distributed. I tried with a course of over 500 students from which 290 were not allocated and the algorithm couldn't finish. It seems that something like in line 222/223 in locallib.php is necessary for larger numbers of students.
2. When restricting a ratingallocate activity to a certain group of users by access restrictions, still all users of the course would be distributed. Of course through the group restriction within choices, the access restriction is technically not necessary but there still could be some use cases where this would be preferable e.g. when doing an allocation for just one group, you could just set the access restriction and would not have to set the group restriction for every choice. But this would lead to problems when distributing not allocated users. The question is how this could be handled. One option would be to no longer allow setting group based access restrictions. However, I'm not sure whether that is such a good idea. A better option would be to also allow this and take only those users for the allocation into account which have access to the activity. What do you think?

PhMemmel commented:
Comment to "3.":
Yes, you are right. I've also stumbled over this already. The thing is: The group feature and the core algorithm also do not respect the distribution of users which are not able to access the activity, but always consider ALL PARTICIPANTS WITH STUDENT ROLE in this course. So I did not target this issue right here.

But I totally agree: The whole activity should only deal with users which can access the activity (and active enrolment BTW). So I would vote to not deal with the groups and the restriction to the activity, but choose a more general approach and make the activity AS A WHOLE only deal with users which have access to the activity (except the activity being invisible of course, in this case distribution for example should work anyway).

PhMemmel commented:
Comment to "1.": I also addressed this. Thank you for testing. I did not realize that it is taking so much resources. The reason might be the following:

When implementing the algorithm first it worked well for small numbers of users. When we first tried to assign many users we found out that we had huge amount of database calls. Distributing in a course with 200 to 300 users took about 18 minutes. So you might have seen that there are a lot of variables including "cached" in the name. This originates from this experience to avoid multiple database calls which were very expensive. After introducing these changes (minimize database calls) the time went down to only a few seconds from about 18 minutes. So caching this data of course comes at the cost of higher memory usage, so I'm not surprised this might be an issue. I added your suggestion.

From my point of view I addressed all the issues you mentioned. Let me know, if you need anything else. Thanks!

Laur0r commented on locallib.php:
```diff
@@ -621,6 +624,259 @@ private function process_action_manual_allocation() {
+        $transaction = $this->db->start_delegated_transaction();
+
+        $usertoassign = array_shift($possibleusers);
+        // As long as we have a user to assign, we try to assign him.
```
The allocation with the 500 students worked however it still took several minutes. I noticed that there were more students than available choices in my course. In this scenario, all students were iterated even when there wasn't any open space at a choice available. This might be a scenario which does not occur regularly but should be considered to avoid unnecessary calculations. I would suggest adding a check how many places are left in general so that the algorithm can directly stop when all spots are filled.

PhMemmel replied:
I had a look at the code and I think I know what you are talking about. I hope I got you right and I added a commit to address this. Thank you very much for this idea.

Also thank you very much doing a test with 500 students. 9 minutes is indeed a long time. It's quite unlikely that 500 students did not vote at all and all of them need to be distributed I'd think, but well...maybe there are teachers/instructors who just use this tool to distribute the students SOMEHOW without even letting them rate first. So, that definitely makes me think if it would be a good idea to run the algorithm as ad hoc task. This of course is not ideal for usability, because the user would press a button and then have to guess when the algorithm has finished, but it's probably better to hit the refresh button a few times than running into a timeout of the reverse proxy or something. Thoughts on this?

Laur0r replied:
Thank you! I have also already thought about running this as an adhoc task. In my opinion, this consideration also applies to the normal allocation algorithm which if not specified differently, is still run without cron. Hence, to be consistent I would leave this as it is and maybe create a new issue for moving everything which could take longer into a cron task.

Laur0r commented:
You are right, the whole activity should not consider all participants in a course but only those students which can access the activity. However for the core algorithm and group feature, this should not make a difference as these are only relevant for the users that can view the activity. Nevertheless, for the manual allocation page it would be great to be only shown users with access and not everyone. This would mean that the function `get_raters_in_course()` is adapted accordingly.

PhMemmel commented:
So you are fine addressing the issue of all users always being considered by mod_ratingallocate in a separate issue?

PhMemmel commented:
I also rebased to current master. I hope I did not mess up solving conflicts :-)

Laur0r commented on locallib.php:
```diff
@@ -842,6 +1087,16 @@ public function handle_view() {
+                $this->distribute_users_without_choice($action);
+                redirect(new moodle_url('/mod/ratingallocate/view.php',
+                            ['id' => $this->coursemodule->id, 'action' => ACTION_NONE]),
+                    get_string('unassigned_users_assigned', ratingallocate_MOD_NAME),
```
This needs to be set to RATINGALLOCATE_MOD_NAME

Laur0r commented on locallib.php:
```diff
@@ -628,6 +631,272 @@ private function process_action_manual_allocation() {
+            if ($choicetoassign === -2) {
+                // This means there are no free places left in any choice for any user, so we can stop the algorithm
+                // as a whole.
+                return;
```
There should be a break here or alternatively add `$transaction->allow_commit()` before the return statement otherwise the made allocations are not registered

Laur0r reviewed (commented):
Thank you for your changes! I found some minor errors resulting from the rebase. Could you also have a look at the failing Codechecker errors?

PhMemmel commented:
@Laur0r I fixed all things you found. Thanks for that.

Regarding the code checker complaints: These are not caused by this PR, but I think because of the rebase? At least codechecker complains about a huge number of errors and warnings. While I am a big fan of green pipelines and very clean code, I would suggest that this should be fixed in a separate codestyle fix PR :-)

Laur0r commented:
@PhMemmel Thank you for your quick response! I certainly agree that the code needs quite a bit of a clean up considering all the Codechecker warnings which are not related to this PR. However, the errors preventing the pipeline from continuing were introduced by your changes (in the `renderer.php` and the test `mod_ratingallocate_allocate_unrated_test.php`). In the master branch the pipeline was able to successfully complete. I would appreciate it if you could adapt those lines so that I can make sure that the behat and phpUnit test still run through.

PhMemmel commented:
Hi @Laur0r,

yes sorry, didn't see that in all that codechecker noise :-)

I tried to fix everything related to my pull request. Hope, everything's good now. Thanks again for checking!

PhMemmel commented:
Failed run seems to be caused by github actions having performance issues right now

PhMemmel commented:
Failed run seems to be caused by github actions having performance issues right now

PhMemmel commented:
Failed run seems to be caused by github actions having performance issues right now

PhMemmel commented:
Failed run seems to be caused by github actions having performance issues right now

PhMemmel commented:
Great news, thank you very much, especially for the very thoroughly review, appreciate it very much!

I just opened the next contribution patch, but please don't feel pressurized :-)
https://github.com/learnweb/moodle-mod_ratingallocate/pull/255
